### ISSUE TITLE
x gantry position unknown when homing xy, designed to be at zmax

### ISSUE THREAD

**Issue by impromptu1583** [mcmillanje/klipperSidekick#1]
Sidekick is designed to park the gantry at the zmax using the TMC2130 safe home feature. The gantry is held at zmax with magnets.

When homing, the gantry will typically be at zmax or zmin. (If the gantry wasn't properly parked it will typically be pulled to zmin by gravity). This introduces a paradox:

If the gantry is at zmin it's not safe to home XY without moving the gantry up a few mm. See [safe z home](https://www.klipper3d.org/BLTouch.html).

However if the gantry is at zmax we don't want to use safe z home and raise it as it's already at its maximum, leading to the steppers grinding against the top of the printer.

**Comment by impromptu1583** [comment]
Options:

1. Assume gantry is at zmin & use safe z home (risk of hitting top of printer if not, but a cleaner solution overall)
2. Assume gantry is at zmax & don't use safe z home (risk of scraping toolhead against bed if not)
3. Implement a software addon to klipper to home zmax first
4. Implement a hardware addon (e.g. physical enstop switch) and macro to home appropriately depending on gantry location

**Comment by cdsmith69** [comment]
I don't believe you can use the bump sense of the Trinamic drivers to home the Z at the top on the Sidekick because it has two Z motors wired in parallel.  I don't know if that interferes with bump sense, or if it is because the two motors won't necessarily hit at the same time.

I think this is also why the Mini 2 and Pro both still use switches to home Z max, even though they use bump sense on the X and Y axes.  But with the Sidekick having a BLTouch, it was possible to just use the BLTouch to find zero at the bottom, then just use a measured move to go to the top.

Because the dimension from the bed to the top might not be exact from one printer to the next is why it slows down shortly before getting to the top.  It then turns the driver current down and moves slowly so when it does hit the top it won't slam hard.

**Comment by impromptu1583** [comment]
That's a really good point. I'll adjust the end-print macro to approach slowly and emulate that behavior. It doesn't solve the issue when initially homing.

Actually maybe it does. I could do a very low
Current z move at the start of xy homing to (hopefully) lift off the bed if the gantry is sitting at z min and see if it can be tuned to not push too hard against zmax if it is there. I'll try it. Thanks!

**Comment by cdsmith69** [comment]
Another thought, maybe since the Sidekick has no Z brake it is safe to assume on power on it is either at the top held by the magnets, or sitting down against the bed.  I wonder if there is a way to use the BLTouch to detect if it powers on in the down position, without moving anything.  Like deploy the BLTouch and see if it goes into an error condition because it can't drop the probe.  Then lift and reset the BLTouch to clear the error.  Don't know if that would work, I don't know a lot about how the BLTouch works.

**Comment by impromptu1583** [comment]
Oh that's an interesting thought. Problem is if the bed is all the way forward (YMax) the bltouch probe is off the end of it. I was hoping to emulate the default behavior (on standard firmware) of "presenting" the print/bed on cancel or finish so there's decent odds the bed would be in that position. Then again if it successfully printed or at least got to the cancel macro it's probably successfully parked the gantry so it does narrow the edge cases significantly.

**Comment by impromptu1583** [comment]
OK I think these are the initial two options I'll pursue (thanks @cdsmith69 for the creative thinking!)

1. a fully stock option that utilizes lower z stepper current to gently try to lift the gantry (will require testing to find the correct value that successfully lifts it from the bed but won't grind forcefully against the top if it's already parked)

2. an option to instead utilize a cheap mechanical endstop mod to detect whether the gantry is parked (which also has potential to solve #4 )

**Comment by impromptu1583** [comment]
Latest commit added 3 options, which can be selected by uncommenting the appropriate line in printer.cfg:

1. low_current_home
z stepper current will be reduced to 15% then move down 10mm, then full current back up 10mm
If nozzle is correctly parked this will result in a quick down-up move pre-homing
If nozzle is resting on bed it won't be able to move down due to low current then will move up (to clear bed for xy home)
this is still bad for your printer and should be avoided (always park the nozzle) but it's better than nothing and doesn't require hardware mod
2. z_stop_home (WIP)
this will require the addition of a physical z endstop (I will make documentation). The homing macro will verify the gantry is properly parked before homing and either notify user or try to park itself if not. This will provide a fast and safe homing and would be the recommended method!
3. danger_home
This will provide a fast homing that skips the failsafes & assumes gantry is parked. If gantry isn't parked **it will slam against the bed and likely cause damage.** Don't use unless you're 100% always going to park your gantry no matter what!
